teps/00XX-interceptor-plugins.md

```diff
@@ -0,0 +1,423 @@
+
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate. Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the WGs or subproject.
+-->
+
+```
+<<[UNRESOLVED @dibyom]>>
+<<[/UNRESOLVED]>>
+```
+
+
+### Performance (optional)
+
+<!--
+Consideration about performance.
+What impact does this change have on the start-up time and execution time
+of task and pipeline runs? What impact does it have on the resource footprint
+of Tekton controllers as well as task and pipeline runs?
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+-->
+
+## Design Details
+
+This section describes the two major portions of the proposal -- the changes to
+the Interceptor interface to make it more typed, and the new CRD to
+install/create interceptors.
+
+### HTTP Interface
+
+This diverges from the current HTTP interface that simply forwards the given request and uses HTTP response codes. Instead the proposed way uses typed response bodies for both requests and response (similar to the AdmissionReviewRequest and Response in AdmissionControllers)
+
+
+#### Request
+
+EL Sink sends incoming events to interceptors over HTTP using a typed format:
+
+
+```go
+	type InterceptorRequest struct {
+		// Body is the incoming HTTP event body
+		Body []byte `json:"body,omitempty"`
+		// Header are the headers for the incoming HTTP event
+		Header map[string][]string `json:"header,omitempty"`
+		// EventUrl is the URL of the incoming event
+		EventUrl string `json:"url,omitempty"`
+		// InterceptorParams are the user specified params for the interceptor
+		InterceptorParams map[string]string `json:"interceptor_ params,omitempty"`
+		// EventID is a unique ID assigned by Triggers to each event
+		EventID string `json:"event_id,omitempty"`
+		// TriggerID is of the form namespace/$ns/triggers/$name
+		TriggerID string `json:"trigger_id,omitempty"`
+	}
+
+```
+
+#### Response
+
+Interceptors in turn return a message with a fixed type:
+
+```go
+type InterceptorResponse struct {
+	// Extensions are  additional fields that is added to the interceptor event.
+	// See TEP-0022. Naming TBD.
+  Extensions       map[string]interface{} `json:"extensions,omitempty"`
+  // Continue indicates if the EventListener should continue processing the Trigger or not
+  Continue         bool                   `json:"continue,omitempty"`
+  // Status is an Error status t 
+	Status           grpc.Status            `json:"status,omitempty"`
+}
+```
+
+### Installing Interceptors
+
+Interceptors are installed via the Interceptor CRD. At minimum the spec declares an `address` field that is the URL where the EventListener should send the request to.
```

Comment:
Not sure if we want to do this (if not, feel free to put this into alternatives), but instead of having users craft the URL, should we have them use an object ref and enforce that they follow the Knative Callable interface?

Reply:
This might help the health checking question below- we can defer to the underlying object status for readiness.

Reply:
Ref: Knative Callable Interface: https://github.com/knative/eventing/blob/master/docs/spec/interfaces.md#callable

Yes, I think that  supporting Callables makes sense. How about supporting both? Just regular URLs or objects that adhere to  the Callable interface?